feat: Support building & using patches as a list of DTOs

acoulton opened the pull request:
Build on the existing implementation to allow users to specify patches as a list of operation objects, rather than as a JSON string.

Unlike the JSON form, the patch operation DTOs are strict about the parameters they accept. If a user wishes to include custom properties, they can implement a class extending the base `PatchOperation`.

Patch DTOs can be serialised to and from JSON, for convenience.

Completes #12 


** note ** I have based this branch off the branch for #13 to avoid merge conflicts. The first 3 commits in this diff can therefore be ignored. Once #13 is merged I will rebase as required.

acoulton committed:
feat: Support building & using patches as a list of DTOs

Build on the existing implementation to allow users to specify patches
as a list of operation objects, rather than as a JSON string.

Unlike the JSON form, the patch operation DTOs are strict about the
parameters they accept. If a user wishes to include custom properties,
they can implement a class extending the base `PatchOperation`.

Patch DTOs can be serialised to and from JSON, for convenience.

acoulton committed:
refactor: Use phpstan type aliases to reference operations from handlers

Now that there are explicit operation classes for each standard
operation, it feels appropriate to define the expected schema for
unserialized operations alongside the DTO and import it to the handler.
This more clearly shows the relationship between these objects.

acoulton commented on src/operations/handlers/AddHandler.php:
```diff
@@ -3,20 +3,18 @@
-     *     path: string,
-     *     value: mixed,
-     * } $patch
+     * @param object&TAddOperationObject $patch
```
Technically, I think the `object&` here is not required - phpstan is happy with just `TAddOperationObject`. 

However, PHPStorm was then complaining that the `@param` does not match the method signature: adding the object intersection type made it happy. 

The other option (that works for phpstan & phpstorm) would be to use `@phpstan-param TAddOperationObject` instead of the `@param` tag.

blancks replied:
This solution seems to be fine with VSCode as well, so I would keep it as is

acoulton commented on src/operations/PatchOperation.php:
```diff
@@ -0,0 +1,10 @@
+
+namespace blancks\JsonPatch\operations;
+
+abstract class PatchOperation
```
The intent is these DTOs will generally only have properties, not methods, and they will always require at least an `op` property. 

Therefore I didn't think there was any benefit in also defining an (empty) interface: users who wish to define a custom operation object should always be able to extend the abstract `PatchOperation` class.

blancks replied:
I agree, for DTO purposes the abstract class is already a strong enough contract

acoulton commented on src/operations/PatchOperationList.php:
```diff
@@ -0,0 +1,77 @@
+                        throw new InvalidPatchOperationException(sprintf('Unknown operation "%s"', $op));
+                    }
+
+                    return new $classes[$op](...$patch);
```
We have no guarantee here that the constructor signature matches the decoded json: extra parameters or incorrect types will cause a TypeError.

For the usecase, I think this is acceptable (and there would be considerable complexity / overhead to attempting to dynamically validate here).

However, a compromise would be to catch the `TypeError` and rethrow it as an `InvalidPatchOperationException` if you'd prefer that?

blancks replied:
It's not an easy task to determine whether a `TypeError` is caused by an invalid patch (in which case catching and rethrowing is the optimal solution, imho) or by an invalid DTO signature (where allowing the TypeError to propagate is the desired behavior).

Since the `FastJsonPatch` class already provides a method to validate a patch beforehand, I wouldn't catch and rethrow here. The current behavior is fine with me.

acoulton commented on tests/operations/PatchOperationListTest.php:
```diff
@@ -0,0 +1,251 @@
+#[UsesClass(Test::class)]
+#[UsesClass(PatchOperation::class)]
+#[UsesClass(FastJsonPatchExceptionTrait::class)]
+#[UsesClass(InvalidPatchException::class)]
+#[UsesClass(InvalidPatchOperationException::class)]
+#[UsesClass(ArrayAccessor::class)]
+#[UsesClass(ArrayAccessorAwareTrait::class)]
+#[UsesClass(ObjectAccessor::class)]
+#[UsesClass(ObjectAccessorAwareTrait::class)]
+#[UsesClass(ValueAccessorAwareTrait::class)]
```
I'm not familiar with running phpunit with the `requireCoverageMetadata` / `beStrictAboutCoverageMetadata` options configured. 

I *think* I've done this correctly (certainly got the coverage build to go green, after a few attempts) although I found it a bit surprising - it feels like the test now has a dependency on implementation details that it shouldn't be aware of. Perhaps that's just the necessary trade-off but let me know if there's something different I should have done.

blancks replied:
Yes, you have! As you guessed, this is a tradeoff to let phpunit know exactly whether a test is intended to cover a specific class. It can feel counterintuitive until you think of it more like an integration test from a visibility perspective. In any case, I've found it very handy in situations like this, where you might be misled into thinking that a test involving multiple classes actually covers them effectively. This is the intended way to use this feature, so there's nothing wrong here.

acoulton commented on tests/FastJsonPatchTest.php:
```diff
@@ -177,6 +193,43 @@ public function testValidJsonPatches(string $json, string $patches, string $expe
+     * @throws \JsonException
+     * @throws FastJsonPatchException
+     */
+    #[DataProvider('validOperationsForDTOsProvider')]
+    public function testValidJsonPatchesAsDTOs(string $json, string $patches, string $expected): void
```
Potentially this test runs more cases than is strictly necessary to prove the DTOs are correctly read & handled.

However it seemed the most efficient way to have confidence that they support all the expected properties / types and work correctly with all handlers and to keep that in sync with any changes to the pure JSON objects over time.

blancks replied:
I'm okay with your assumptions here, but I would move the tests into a dedicated class for each operation DTO. I suggest this because I think each DTO also needs a test that ensures it fails when incorrect properties are passed. This is important because the `PatchOperationList::fromJson` method bases its validation workflow (both for patch validation and invalid DTOs) on the assumption that the constructor throws an exception when invalid input is provided.

acoulton replied:
I'm not sure about creating separate test classes for the individual DTOs - they have no logic of their own, only a constructor and some promoted properties. I wouldn't usually unit test things like that other than through the code that uses them - I think for example we can trust that PHP will handle a user doing something explicitly wrong like `new Add()` or `new Add(op: new DateTimeImmutable)` without adding a test to show that?

Also just to be clear I wasn't imagining the DTOs would be validated as such (other than basic type-safety / property naming). For example PHP will natively prevent you creating a DTO with an `array` for the `path`, but you could pass any string even if it's not a valid JSON pointer. Actual validation only happens if you pass the DTO to FastJsonPatch. Otherwise the DTOs start to have dependencies like the `jsonPointerHandler` and stop being DTOs.

However you're right that I'm missing examples of attempting to build DTOs from invalid JSON (e.g. the items in the JSON list aren't patch objects / are missing properties / have extra properties / have properties with the wrong types). Even if we're ok with PHP causing native TypeError there, we should prove that's what happens. I think I should just add those to the cases in `PatchOperationListTest::invalidJsonProvider` covering `PatchOperationList::fromJson()` which is the only place I'm dynamically creating DTOs.

Can you let me know if you're happy with that and I'll update?

blancks replied:
I wasn't implying adding validation logic to the DTOs, but since we're relying on their types as a safety mechanism, we need to prove that they're type-safe and not constructed with a mixed argument, for example. 

I suggested separate test classes mainly because it felt cleaner, but I understand your point, and I'm fine testing this the way you proposed as long as we agree that the behaviour should be covered

acoulton replied:
I have added more cases to the `PatchOperationListTest` and marked that that class now also covers the individual DTOs (since it contains examples of creating every supported DTO with valid values, and invalid values for a sample of them).

In implementing this, I found a few other actual or theoretical cases where an invalid data structure in the JSON would cause a runtime error (e.g. `op` is not specified or is an invalid type to be an array key). I decided it was more robust to check for these and throw an appropriate `InvalidPatchException` / `InvalidPatchOperationException`.

That led me to also catching and wrapping exceptions when creating the actual DTOs. Having thought about it some more, I think this is cleaner and fairly robust - if we assume that the DTO has tests proving it can be created with valid values (ours do) then any TypeError / ArgumentCountError / incorrect named parameters errors at runtime are virtually guaranteed to be due to invalid values in the JSON.

Of course, if a user gets an `InvalidPatchOperationException` for something they think should be valid, they can open a bug (as they would if it was our own validation logic that was incorrect).

I'm not sure about your comment `we need to prove they're type-safe and not constructed with a mixed argument, for example`. 

We can see that the individual DTOs are typed with `string` where they should be from their class definitions.

We could add a test for every individual DTO to prove that e.g. we get an exception with `{... "path": true}` - but that could still pass if the property type was changed to `string|int`. The only way to conclusively prove the DTO will *only* accept a `string` would be to have test cases for every parameter with every possible type of non-string value. That feels redundant to me given none of those cases depend on any logic of our own.

I also saw that the existing tests for the library don't go into that level of detail testing invalid property data types in the native `stdClass` unserialisation so I think I might be misunderstanding what you want here?

I *think* all we need to test is that a suitable exception is thrown if **any** DTO is created with **any** parameter of an incorrect type.

Therefore I've just added a test case for `'[{"op":"add", "path": true, "value": "bar"}]'` which IMO covers all cases where one or more parameters don't match the type declared in the DTO constructor.

Let me know what you think of this?

blancks replied:
This all looks good to me 👍

To clarify my earlier point: I wasn't aiming for exhaustive testing of PHP's type system or suggesting validation logic in the DTOs themselves. My concern was really just about making the assumption explicit and covered by tests: that invalid JSON results in a failure at DTO construction time, since PatchOperationList::fromJson() relies on that behaviour for its validation flow.

The additional invalid JSON cases you've added cover that expectation well, and I'm happy with testing this through PatchOperationListTest rather than separate DTO test classes. Wrapping the underlying runtime errors into domain-specific exceptions also makes the behaviour clearer and more robust from a consumer's point of view.

So overall I'm comfortable with the current approach and coverage. 
Thanks for thinking this through!

blancks commented on src/operations/Add.php:
```diff
@@ -2,76 +2,19 @@
-
-        return ['op' => 'replace', 'path' => $patch->path, 'value' => $this->previous];
+    public function __construct(
+        public readonly string $path,
```
Since PHP 8.1 is reaching EOL on January 1st, I'm considering bumping the minimum required PHP version to 8.2 for the 3.0 release. This would allow us to use the class-level `readonly` keyword.

What do you think about that?

acoulton replied:
That sounds a good idea - I'm certainly fine with that personally, everything I would want to use this with is already on 8.4 or later.

acoulton replied:
I've updated the DTOs to be `readonly` classes, except for the abstract `PatchOperation` base class - if we're using that as the contract for users to add their own implementations, they should probably be able to decide whether to make them readonly or not? (the `op` property in `PatchOperation` is still defined as readonly though as a basic requirement).

I wasn't sure if you'd want to bump the PHP version in 2.x or 3.x - happy to do a PR to update the composer.json and github actions etc if you let me know whether to base off `main` or `dev`?

blancks replied:
@acoulton I have created and merged PR #15, which bumps the min php version to 8.2 for the existing actions. You should be able to merge the updated `dev` branch into your branch without conflicts.

This will be part of the 3.0 release as well

acoulton replied:
Thanks :)

I've rebased onto `dev`. Of course, the `PatchOperation` base class also has to be readonly as a readonly class cannot extend a non-readonly class. On reflection, I think that's OK - it seems reasonable to enforce that end-user DTO classes must also be readonly.

blancks commented on src/operations/PatchOperationList.php:
```diff
@@ -0,0 +1,77 @@
+        JsonHandlerInterface $jsonHandler = new BasicJsonHandler(),
+        array $customClasses = [],
+    ): self {
+        $patches = $jsonHandler->decode($jsonOperations, ['associative' => true]);
```
Some JSON decoders may not support the associative array option at all because forcing objects into arrays introduces representation limitations that make it error-prone. Decoding the patch as objects instead is a more robust approach.

acoulton replied:
Of course, you're absolutely right. I hadn't quite thought through that although the top-level patches will always decode OK as an array, the `value` property is completely arbitrary and could easily differ between `[]` and `{}`

acoulton replied:
Fixed now and I've added a test case explicitly using an object `value` to prove that it is encoded and decoded to JSON as expected.

acoulton committed:
fix: PatchOperationList should decode patches as objects

Because although the top-level patch entries will be equivalent as
arrays or objects, properties like the `value` key can contain any type
of data.

acoulton committed:
feat: Make all `final` patch operation classes `readonly`

Will break on 8.1, to be rebased once the project drops 8.1 support.

I have intentionally left the base `PatchOperation` class as *not*
readonly in case end-users want to extend it with mutable operation
DTOs for any reason. Instead, I have just left `op` as a readonly
property in the base class.

acoulton committed:
feat: More strictly check patch structural validity during fromJson

Improve test coverage of attempting to create a PatchOperationList from
JSON with invalid data. Update the implementation so that the majority
of potential issues will be detected and thrown as InvalidPatchException
or InvalidPatchOperationException.

acoulton committed:
test: Prove that PatchOperationList::fromJson bubbles unexpected errors

It's essentially impossible for our own DTOs to trigger PHP errors other
than the ones related to creating them with incorrect parameters.

However, because our code has to catch `Error` to check for certain
types of error, we need to prove that it rethrows if the error is not
one that it expects to handle.

acoulton commented on src/operations/PatchOperationList.php:
```diff
@@ -0,0 +1,155 @@
+     */
+    private static function createPatchDtoFromValues(string $op, string $class, array $values): PatchOperation
+    {
+        try {
+            return new $class(...$values);
+        } catch (ArgumentCountError $e) {
```
I have gone for catching & rethrowing (even though it is a little verbose) because it minimises runtime performance impact in the "happy case" which is likely the most common.

Alternatively we could use Reflection to get the list of expected constructor parameter names and types and strictly compare that to the `$values` array before we attempt to create the object. 

However this would create a performance overhead and would likely need to be cached (at least in process, perhaps between processes). That would complicate the implementation & likely introduce dependencies.

Given we know that the objects are expected to be very simple value objects, that will almost always have the right data (and that are properly validated when they get to FastJsonPatch and the relevant handler) I think this is an OK solution?

blancks replied:
Yes, this works for me 👍

acoulton commented on tests/operations/PatchOperationListTest.php:
```diff
@@ -0,0 +1,369 @@
+    }
+}
+
+readonly class Append extends PatchOperation
```
Unfortunately anonymous classes cannot be `readonly` until PHP 8.3, so the custom classes used by these tests have to be defined as actual named classes.

acoulton commented:
One other thought: I originally only wrote the code for `PatchOperationList::fromJson` as a helper method within a testcase to allow me to use the existing json test data for testing `FastJsonPatch` with the DTOs.

Then I thought it might be useful in the library itself so moved it into that public static constructor.

If you're not sure about the testing/robustness of that method and the validation etc I could easily move it back to being an internal test helper. 

If we wanted to go down the route of properly checking things (e.g. with Reflection) it would be cleaner to do that with an instantiable hydrator class which could then have dependencies. That could be added later as a separate PR.

blancks commented:
@acoulton first of all, thank you for the amazing effort you're putting into this project and sorry that I'm only getting to review this now.

The helper method is actually a nice addition to the library, and I'm happy with it being public and used in tests as well.

Going down the reflection route feels a bit redundant to me. Letting DTO instantiation fail naturally is good enough in this case, especially now that we have tests explicitly covering that behaviour. If we ever need stricter or more explicit validation later on, introducing a dedicated hydrator as a separate concern would make sense, but I don't think that's necessary right now.

blancks commented on src/FastJsonPatch.php:
@acoulton I would add the `PatchOperationList` type to the `isValidPatch` method here as well. 

This should be the very last thing before merging the PR 😃

acoulton replied:
Thanks @blancks and no need to apologise for the review timing :)

That's a good suggestion, I've added that now.

acoulton committed:
feat: `isValidPatch` should also accept `PatchOperationList`

Internally, this is already implemented - it just required a change to
the typehint for `isValidPatch`.

I've refactored the existing `isValidPatch` tests to use a DataProvider
to make the overlap between examples for strings and DTOs clearer. As
part of this I added a couple of new examples of valid / invalid string
patches. This is because some of the existing string cases (missing
params / structural issues) cannot happen with DTOs.

acoulton commented on tests/FastJsonPatchTest.php:
```diff
@@ -58,25 +66,74 @@
+     */
+    public static function isValidPatchProvider(): array
     {
-        $FastJsonPatch = FastJsonPatch::fromJson('{"foo":"bar"}');
-        $this->assertFalse($FastJsonPatch->isValidPatch('{"op":"test","path":"/foo","value":"bar"}'));
-        $this->assertFalse($FastJsonPatch->isValidPatch('[{"op":"add"}]'));
```
I refactored the existing `isValidPatch` tests to use a data provider, to make it easier to see the overlap / differences between the string and DTO examples that are tested.

Some of the previous string examples (structural issues / missing parameters) cannot be represented as DTOs. So I also added a couple of new string examples to prove that these behave consistently with both a string patch and a DTO.

acoulton commented on tests/FastJsonPatchTest.php:
```diff
@@ -58,25 +66,74 @@
+                new PatchOperationList(new Test(path: '/nonexistent-path', value: 'any')),
+                true,
+            ],
+            'DTO patch - invalid (invalid path)' => [
+                new PatchOperationList(new Remove(path: 'not a path')),
+                false,
+            ],
```
This (invalid JSON pointer string) was actually the only case I could think of where a DTO created in code could actually fail validation.
